File: packages/hyperfly/src/http.ts
import type { Codec } from "./codec.js";
import type { CodecRegistry } from "./registry.js";

export const HYPERFLY_MEDIA_TYPE = "application/vnd.hyperfly";
export const ACCEPT_HEADER = "hyperfly-accept";
export const CODEC_HEADER = "hyperfly-codec";
export const OFFER_HEADER = "hyperfly-offer";
export const WELL_KNOWN_PREFIX = "/.well-known/hyperfly/";

/** Negotiation §6: client-controlled input, so parsing is bounded. */
const MAX_ACCEPTED = 32;
const FINGERPRINT = /^[0-9a-f]{32}$/;

/**
 * Fingerprints the client says it can decode, in its order of preference.
 * Malformed entries are dropped rather than failing the request.
 */
export function parseAccept(header: string | null | undefined): string[] {
  if (!header) return [];
  const out: string[] = [];
  for (const part of header.split(",")) {
    if (out.length >= MAX_ACCEPTED) break;
    const value = part.trim().toLowerCase();
    if (FINGERPRINT.test(value) && !out.includes(value)) out.push(value);
  }
  return out;
}

export type Negotiation<T = unknown> =
  | { kind: "hyperfly"; codec: Codec<T>; headers: Record<string, string> }
  | { kind: "json"; headers: Record<string, string> };

export interface NegotiateOptions {
  /** Named in Hyperfly-Offer when falling back, so a client can upgrade itself. */
  offer?: string;
  /** An operator switch or a load shed: fall back without consulting the registry. */
  enabled?: boolean;
}

/**
 * Decide how to answer one request. Framework-agnostic on purpose: everything a
 * server needs is the Hyperfly-Accept header and a registry.
 *
 * Vary is always set, because the same URL yields either representation and a
 * shared cache would otherwise hand one peer's binary to a peer that cannot read it.
 */
export function negotiate<T = unknown>(
  accept: string | null | undefined,
  registry: CodecRegistry,
  options: NegotiateOptions = {},
): Negotiation<T> {
  const vary = { Vary: "Hyperfly-Accept" };

  if (options.enabled === false) {
    return { kind: "json", headers: { ...vary, "Content-Type": "application/json" } };
  }

  const codec = registry.select(parseAccept(accept));
  if (codec) {
    return {
      kind: "hyperfly",
      codec: codec as unknown as Codec<T>,
      headers: {
        ...vary,
        "Content-Type": HYPERFLY_MEDIA_TYPE,
        "Hyperfly-Codec": codec.fingerprint,
      },
    };
  }

  const offer = options.offer ?? registry.fingerprints[0];
  return {
    kind: "json",
    headers: {
      ...vary,
      "Content-Type": "application/json",
      ...(offer ? { "Hyperfly-Offer": offer } : {}),
    },
  };
}

export interface Encoded {
  body: Uint8Array | string;
  headers: Record<string, string>;
}

/** Encode one value according to a negotiation decision. */
export function encodeFor<T>(decision: Negotiation<T>, value: T): Encoded {
  if (decision.kind === "hyperfly") {
    return { body: decision.codec.encode(value), headers: decision.headers };
  }
  return { body: JSON.stringify(value), headers: decision.headers };
}

export interface WellKnownResponse {
  status: number;
  body: string;
  headers: Record<string, string>;
}

/**
 * Serve `.well-known/hyperfly/{fingerprint}` (negotiation §3). Artifacts are
 * content-addressed, so a hit is immutable and cacheable forever; a miss is a 404
 * and not an error — the client simply stays on JSON.
 */
export function serveArtifact(pathname: string, registry: CodecRegistry): WellKnownResponse | undefined {
  if (!pathname.startsWith(WELL_KNOWN_PREFIX)) return undefined;
  const fingerprint = pathname.slice(WELL_KNOWN_PREFIX.length).toLowerCase();
  if (!FINGERPRINT.test(fingerprint)) {
    return { status: 404, body: "", headers: { "Cache-Control": "no-store" } };
  }
  const artifact = registry.artifact(fingerprint);
  if (!artifact) {
    return { status: 404, body: "", headers: { "Cache-Control": "no-store" } };
  }
  return {
    status: 200,
    body: artifact,
    headers: {
      "Content-Type": "application/json",
      "Cache-Control": "public, max-age=31536000, immutable",
    },
  };
}

/** The header a client sends. Preference order is the caller's. */
export function acceptHeader(fingerprints: readonly string[]): string {
  return fingerprints.join(", ");
}

/**
 * Decode a response body according to what the server said it sent. A codec the
 * client does not hold is not an error the client can recover from by guessing,
 * so it is reported as a miss and the caller re-requests as JSON.
 */
export function decodeResponse<T>(
  contentType: string | null | undefined,
  body: Uint8Array | string,
  registry: CodecRegistry,
): { kind: "json"; value: T } | { kind: "hyperfly"; value: T } | { kind: "unknown-codec"; fingerprint: string } {
  const isHyperfly = (contentType ?? "").toLowerCase().startsWith(HYPERFLY_MEDIA_TYPE);
  if (!isHyperfly) {
    const text = typeof body === "string" ? body : new TextDecoder().decode(body);
    return { kind: "json", value: JSON.parse(text) as T };
  }
  const bytes = typeof body === "string" ? new TextEncoder().encode(body) : body;
  let fingerprint = "";
  for (let i = 3; i < 19 && i < bytes.length; i++) fingerprint += bytes[i]!.toString(16).padStart(2, "0");
  const codec = registry.get(fingerprint);
  if (!codec) return { kind: "unknown-codec", fingerprint };
  return { kind: "hyperfly", value: codec.decode(bytes) as T };
}

/**
 * Fetch-API glue. One function covers Hono, Cloudflare Workers, Bun.serve, Deno and
 * Next route handlers, because they all speak Request and Response.
 */
export interface FetchHandlerOptions extends NegotiateOptions {
  /** Also answer .well-known artifact discovery. On by default (negotiation §3). */
  discovery?: boolean;
  status?: number;
  headers?: Record<string, string>;
}

/** Answer one request with `value`, in binary when the client can read it. */
export function respond<T>(
  request: Request,
  value: T,
  registry: CodecRegistry,
  options: FetchHandlerOptions = {},
): Response {
  const decision = negotiate<T>(request.headers.get(ACCEPT_HEADER), registry, options);
  const { body, headers } = encodeFor(decision, value);
  return new Response(body as BodyInit, {
    status: options.status ?? 200,
    headers: { ...headers, ...options.headers },
  });
}

/**
 * Artifact discovery as a Response, or undefined when the path is not ours — so a
 * caller can chain it ahead of its own router in one line.
 */
export function discovery(request: Request, registry: CodecRegistry): Response | undefined {
  const served = serveArtifact(new URL(request.url).pathname, registry);
  if (!served) return undefined;
  return new Response(served.body || null, { status: served.status, headers: served.headers });
}

/** Read a hyperfly or JSON request body (negotiation §4). */
export async function readBody<T>(
  request: Request,
  registry: CodecRegistry,
): Promise<{ ok: true; value: T } | { ok: false; response: Response }> {
  const contentType = request.headers.get("content-type");
  if (!(contentType ?? "").toLowerCase().startsWith(HYPERFLY_MEDIA_TYPE)) {
    return { ok: true, value: (await request.json()) as T };
  }
  const bytes = new Uint8Array(await request.arrayBuffer());
  const decoded = decodeResponse<T>(contentType, bytes, registry);
  if (decoded.kind === "unknown-codec") {
    // a body already sent in a format we cannot read has no safe fallback
    return {
      ok: false,
      response: new Response(null, {
        status: 415,
        headers: {
          ...(registry.fingerprints[0] ? { "Hyperfly-Offer": registry.fingerprints[0] } : {}),
        },
      }),
    };
  }
  return { ok: true, value: decoded.value };
}

File: python/src/hyperfly/http.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from typing import Any

from ._codec import Codec
from .registry import CodecRegistry

HYPERFLY_MEDIA_TYPE = "application/vnd.hyperfly"
ACCEPT_HEADER = "hyperfly-accept"
CODEC_HEADER = "hyperfly-codec"
OFFER_HEADER = "hyperfly-offer"
WELL_KNOWN_PREFIX = "/.well-known/hyperfly/"

_MAX_ACCEPTED = 32
_FINGERPRINT = re.compile(r"^[0-9a-f]{32}$")


def parse_accept(header: str | None) -> list[str]:
    """Fingerprints the client can decode, in its order of preference.

    Client-controlled input, so parsing is bounded and malformed entries are dropped
    rather than failing the request (negotiation section 6).
    """
    if not header:
        return []
    out: list[str] = []
    for part in header.split(","):
        if len(out) >= _MAX_ACCEPTED:
            break
        value = part.strip().lower()
        if _FINGERPRINT.match(value) and value not in out:
            out.append(value)
    return out


@dataclass(frozen=True)
class Negotiation:
    kind: str
    headers: dict[str, str]
    codec: Codec | None = None


def negotiate(
    accept: str | None,
    registry: CodecRegistry,
    *,
    offer: str | None = None,
    enabled: bool = True,
) -> Negotiation:
    """Decide how to answer one request.

    Vary is always set: the same URL yields either representation, and a shared cache
    would otherwise hand one peer's binary to a peer that cannot read it.
    """
    vary = {"Vary": "Hyperfly-Accept"}

    if not enabled:
        return Negotiation("json", {**vary, "Content-Type": "application/json"})

    codec = registry.select(parse_accept(accept))
    if codec is not None:
        return Negotiation(
            "hyperfly",
            {**vary, "Content-Type": HYPERFLY_MEDIA_TYPE, "Hyperfly-Codec": codec.fingerprint},
            codec,
        )

    chosen = offer or (registry.fingerprints[0] if registry.fingerprints else None)
    headers = {**vary, "Content-Type": "application/json"}
    if chosen:
        headers["Hyperfly-Offer"] = chosen
    return Negotiation("json", headers)


def encode_for(decision: Negotiation, value: Any) -> tuple[bytes, dict[str, str]]:
    if decision.kind == "hyperfly" and decision.codec is not None:
        return decision.codec.encode(value), decision.headers
    return json.dumps(value, separators=(",", ":")).encode("utf-8"), decision.headers


@dataclass(frozen=True)
class ArtifactResponse:
    status: int
    body: str
    headers: dict[str, str] = field(default_factory=dict)


def serve_artifact(pathname: str, registry: CodecRegistry) -> ArtifactResponse | None:
    """Serve .well-known artifact discovery, or None when the path is not ours.

    Artifacts are content-addressed, so a hit is immutable and cacheable forever; a
    miss is a 404 and not an error, because the client simply stays on JSON.
    """
    if not pathname.startswith(WELL_KNOWN_PREFIX):
        return None
    fingerprint = pathname[len(WELL_KNOWN_PREFIX) :].lower()
    if not _FINGERPRINT.match(fingerprint):
        return ArtifactResponse(404, "", {"Cache-Control": "no-store"})
    artifact = registry.artifact(fingerprint)
    if artifact is None:
        return ArtifactResponse(404, "", {"Cache-Control": "no-store"})
    return ArtifactResponse(
        200,
        artifact,
        {"Content-Type": "application/json", "Cache-Control": "public, max-age=31536000, immutable"},
    )


def accept_header(fingerprints: list[str]) -> str:
    return ", ".join(fingerprints)


def decode_response(content_type: str | None, body: bytes | str, registry: CodecRegistry) -> tuple[str, Any]:
    """Returns (kind, value) where kind is 'json', 'hyperfly', or 'unknown-codec'.

    A codec the client does not hold is not something it can recover from by guessing,
    so it is reported rather than attempted.
    """
    if not (content_type or "").lower().startswith(HYPERFLY_MEDIA_TYPE):
        text = body.decode("utf-8") if isinstance(body, bytes) else body
        return "json", json.loads(text)
    data = body.encode("utf-8") if isinstance(body, str) else body
    fingerprint = data[3:19].hex()
    codec = registry.get(fingerprint)
    if codec is None:
        return "unknown-codec", fingerprint
    return "hyperfly", codec.decode(data)
